src/mqtt.c

```diff
@@ -526,20 +537,188 @@ static int mqtt_write(const data_set_t *ds, const value_list_t *vl,
     return status;
   }
 
-  status = format_values(payload, sizeof(payload), ds, vl, conf->store_rates);
+  if (conf->format == MQTT_FORMAT_JSON) {
+    format_json_initialize(payload, &offset, &bfree);
+    format_json_value_list(payload, &offset, &bfree, ds, vl, conf->store_rates);
+    status = format_json_finalize(payload, &offset, &bfree);
+
+    if (status != 0) {
+      ERROR("mqtt plugin: format_json_finalize failed with status %d.", status);
+      return status;
+    }
+  } else { /* MQTT_FORMAT_PLAIN */
+    status = format_values(payload, sizeof(payload), ds, vl, conf->store_rates);
+    if (status != 0) {
+      ERROR("mqtt plugin: format_values failed with status %d.", status);
+      return status;
+    }
+  }
+
+  status = publish(conf, topic, payload, strlen(payload));
+  if (status != 0) {
+    ERROR("mqtt plugin: publish failed: %s", mosquitto_strerror(status));
+    return status;
+  }
+
+  return status;
+} /* mqtt_write */
+
+static int format_notification_topic(char *buf, size_t buf_len,
+                                     notification_t const *n,
+                                     mqtt_client_conf_t *conf) {
+  char name[MQTT_MAX_TOPIC_SIZE];
+  int status;
+  char *c;
+
+  if ((conf->notification_prefix == NULL) ||
+      (conf->notification_prefix[0] == 0)) {
+    // tempting but unsafe to use FORMAT_VL here
+    return format_name(buf, buf_len, n->host, n->plugin, n->plugin_instance,
+                       n->type, n->type_instance);
+  }
+
+  status = format_name(name, sizeof(name), n->host, n->plugin,
+                       n->plugin_instance, n->type, n->type_instance);
+  if (status != 0)
+    return status;
+
+  status = ssnprintf(buf, buf_len, "%s/%s", conf->notification_prefix, name);
+  if ((status < 0) || (((size_t)status) >= buf_len))
+    return ENOMEM;
+
+  while ((c = strchr(buf, '#')) || (c = strchr(buf, '+'))) {
+    *c = '_';
+  }
+
+  return 0;
+} /* format_notification_topic */
+
+static int format_plain_notification(char *buffer, size_t buffer_size,
```

Comment:
For consistency with other functions:
```suggestion
static int format_plain_notification(char *buf, size_t buf_len,
```

Reply:
Done!

I note, for future refactorings...

```
$ ack --cc 'buffer_size'  src | wc -l
507
$ ack --cc 'buf_len'  src | wc -l
85
$ ack --cc 'buf_size'  src | wc -l
58
```

Reply:
Thanks, I did not check rest of collectd.  While it would be good to be consistent within whole project, IMHO it's more important to be consistent with same plugin (I think most people deal only with a single one).